src/lib.rs

```diff
@@ -35,18 +35,24 @@ pub type SessionId = u64;
 
 /// The source of data items that consensus should order.
 ///
-/// AlephBFT internally calls [`DataIO::get_data`] whenever a new unit is created and data needs to be placed inside.
-/// The [`DataIO::send_ordered_batch`] method is called whenever a new round has been decided and thus a new batch of units
-/// (or more precisely the data they carry) is available.
+/// AlephBFT internally calls [`DataProvider::get_data`] whenever a new unit is created and data needs to be placed inside.
 ///
 /// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/aleph_bft_api.html for a discussion
 /// and examples of how this trait can be implemented.
-pub trait DataIO<Data> {
-    type Error: Debug + 'static;
+pub trait DataProvider<Data> {
     /// Outputs a new data item to be ordered
     fn get_data(&self) -> Data;
-    /// Takes a new ordered batch of data item.
-    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+/// The source of finalization of the units that consensus produces.
+///
+/// The [`FinalizationProvider::data_finalized`] method is called whenever a new unit
+/// (or more precisely the data they carry) is available.
+pub trait FinalizationProvider<Data> {
```

Comment:
I'm a bit bothered by the naming – both these traits are called `Provider`, but one provides data to the algorithm, while the other handles pieces of data that have been finalized by it.

Renaming this to `FinalizationHandler` would be better imo, but there might also be an even better naming schema possible here.

Reply:
I didnt like this name either. Handler look fine to me, what about `FinalizationNotificator`?

Reply:
Nevermind I dont like this name anymore :P

Reply:
Ok, done